Implements the first half of dragway::RoadGeometry::DoToRoadPosition().

liangfok opened the pull request:
This implements `dragway::RoadGeometry::DoToRoadPosition()` for the scenario where the projection of the provided `GeoPosition` in the x-y plane is on the driveable region of the dragway. A followup PR will implement the case where the provided `GeoPosition` is not in the driveable region of the dragway.

Partially resolves #5269.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5313)
<!-- Reviewable:end -->

liangfok commented:
+@maddog-tri for feature review please.
+@ggould-tri for platform review please.
+@jwnimmer-tri for car review please.

Thanks!

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-Kdm3L_5PSwF4xqRT3P8:brs22nr)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/road_geometry.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-KdmDL2SDHgv4jVbP8S4:-KdmDL2SDHgv4jVbP8S5:b-oa2b9j) ([raw file](https://github.com/robotlocomotion/drake/blob/10bd6178ef76c27c9c8ba640a429e28fae0e7e3d/drake/automotive/maliput/dragway/road_geometry.cc#L66)):*
> ```cpp
>         "dragway::RoadGeometry::IsGeoPositionOnDragway(): The provided geo_pos "
>         "({}, {}) is not on the dragway because its x value is less than zero.",
>         geo_pos.x, geo_pos.y);
> ```

Text logging for geometry queries sounds like an awful plan.   These queries (and thus logs) could happen at multiples of the tick rate!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> automotive design review

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-KdmGZBBpR6dH5nnNa3z:b8xgnlq)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-KdmDL2SDHgv4jVbP8S4:-KdmRBkruDQ45Sy0rk3_:b-ridrnm) ([raw file](https://github.com/robotlocomotion/drake/blob/10bd6178ef76c27c9c8ba640a429e28fae0e7e3d/drake/automotive/maliput/dragway/road_geometry.cc#L66)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Text logging for geometry queries sounds like an awful plan.   These queries (and thus logs) could happen at multiples of the tick rate!
</blockquote></details>

Ah, good point. How about I downgrade these to trace?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/road_geometry.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-KdmDL2SDHgv4jVbP8S4:-KdmdosSwwJVyJTZJlS_:b-vzejgx) ([raw file](https://github.com/robotlocomotion/drake/blob/10bd6178ef76c27c9c8ba640a429e28fae0e7e3d/drake/automotive/maliput/dragway/road_geometry.cc#L66)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Ah, good point. How about I downgrade these to trace?
</blockquote></details>

Trace would be tolerable with a clothespin on my nose (I would not block a merge because of it).  But if we think trace logging is needed, wouldn't "if (!result) { log length,maxy,miny }" at the end of the funtion be much more brief?  I think we can trust readers to run the comparisons in their head without spelling it out four times.

But really, details like the logs below what debuggers are for.  And if someone can't figure out whether a 2d point lies in an 2s axis-aligned rectangle without all of this logging, I'm not sure that they will be able to figure out how to put the logger into trace mode either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-KdmDL2SDHgv4jVbP8S4:-Kdne39Ueo83pmiwjojx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/10bd6178ef76c27c9c8ba640a429e28fae0e7e3d/drake/automotive/maliput/dragway/road_geometry.cc#L66)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Trace would be tolerable with a clothespin on my nose (I would not block a merge because of it).  But if we think trace logging is needed, wouldn't "if (!result) { log length,maxy,miny }" at the end of the funtion be much more brief?  I think we can trust readers to run the comparisons in their head without spelling it out four times.

But really, details like the logs below what debuggers are for.  And if someone can't figure out whether a 2d point lies in an 2s axis-aligned rectangle without all of this logging, I'm not sure that they will be able to figure out how to put the logger into trace mode either.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
-@maddog-tri from feature review.
+@jadecastro for feature review please.

Thanks!

---

Review status: 1 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-KdxUcWh_yB_7JDiBLOK:b-1qxehy)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-KdzuIdbYXlCx-xtCh_k:-KdzuIdbYXlCx-xtCh_l:bnj7jd) ([raw file](https://github.com/robotlocomotion/drake/blob/41012fea4d40307be665e822fdd3713f6af583e1/drake/automotive/maliput/dragway/road_geometry.cc#L66)):*
> ```cpp
>     result = false;
>   }
> ```

FYI -- Setting `result`, ending the if block, and then immediately testing for `result` is a bit clumsy.  Did you write it this way because you mean to add lots more result-setting blocks in the future?  In any case you could factor out the second `if`, remove the result temporary in favor of an immediate return, and end up with a shorter and clearer method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-KdzuIdbYXlCx-xtCh_k:-Kdzxcw8C6NV-N2KtZpS:b-lxirt8) ([raw file](https://github.com/robotlocomotion/drake/blob/41012fea4d40307be665e822fdd3713f6af583e1/drake/automotive/maliput/dragway/road_geometry.cc#L66)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- Setting `result`, ending the if block, and then immediately testing for `result` is a bit clumsy.  Did you write it this way because you mean to add lots more result-setting blocks in the future?  In any case you could factor out the second `if`, remove the result temporary in favor of an immediate return, and end up with a shorter and clearer method.
</blockquote></details>

Oops. `r1` of this PR had a much verbose set of conditionals with numerous separate `if` statements. Now that we've collapsed it all into a single `if` statement, you're right. I've refactored the code as suggested.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 66 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-KdzuIdbYXlCx-xtCh_k:-KdzzR0jBmbalUu1W_T0:b32vhw) ([raw file](https://github.com/robotlocomotion/drake/blob/41012fea4d40307be665e822fdd3713f6af583e1/drake/automotive/maliput/dragway/road_geometry.cc#L66)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Oops. `r1` of this PR had a much verbose set of conditionals with numerous separate `if` statements. Now that we've collapsed it all into a single `if` statement, you're right. I've refactored the code as suggested.
</blockquote></details>

:)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-KdzzVF-pEBwAp6x7M_7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Pass complete. PTAL.

---

Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke1TwrtvZZ55Glk52sQ:-Ke1TwrtvZZ55Glk52sR:b-pnczjx) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/road_geometry.cc#L52)):*
> ```cpp
> }
> 
> bool RoadGeometry::IsGeoPositionOnDragway(const api::GeoPosition& geo_pos)
> ```

There doesn't seem to be any unit test coverage for this function. In particular, you might want to consider adding a test that checks the validity of the `false` result.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 297 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke24-lmPUOU42u3BlGE:-Ke24-lmPUOU42u3BlGF:b-li7zf9) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L297)):*
> ```cpp
> 
>   // Spot checks geographic positions on lane 0 and the right shoulder.
>   for (double x = 0; x <= 100; x += 0.5) {
> ```

BTW, consider replacing the limits on `x`, `y`, and `z` with their named variables `length_`, `lane_width_`, etc.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke27PFIN6VmUVjKxcBW:bhm6r8e) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
> ```cpp
>   for (double x = 0; x <= 100; x += 0.5) {
>     for (double y = -6.5; y <= 0; y += 0.1) {
>       for (double z = 0; z <= 10; z += 2.5) {
> ```

BTW, there are over 65,000 points checked in these two unit tests alone. Is the entire grid necessary? Or is it possible to pare this down to a handful of interesting points you'd like to check?

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 320 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke24ax6NevJzPfBPDhu:-Ke24ax6NevJzPfBPDhv:b-li7zf9) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L320)):*
> ```cpp
> 
>   // Spot checks geographic positions on lane 1 and the left shoulder.
>   for (double x = 0; x <= 100; x += 0.5) {
> ```

BTW, consider replacing the limits on `x`, `y`, and `z` with their named variables `length_`, `lane_width_`, etc.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 338 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke22ndSOQ9uWGxnOeuD:-Ke22ndSOQ9uWGxnOeuE:b-z8u5i2) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L338)):*
> ```cpp
>         EXPECT_EQ(road_position.pos.s, x);
>         if (y == 0) {
>           EXPECT_EQ(road_position.pos.r, y + 3);
> ```

BTW, consider replacing `3`  with `lane_width / 2`.  Ditto below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-Ke272ibVBKJ7PshdnNp:b-wjpjkm)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke1TwrtvZZ55Glk52sQ:-Ke3_fjg1FbWyN6GLJ96:b-g4tvri) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/road_geometry.cc#L52)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

There doesn't seem to be any unit test coverage for this function. In particular, you might want to consider adding a test that checks the validity of the `false` result.
</blockquote></details>

Since this is a private method, the only way to unit test this is to check whether `ToRoadPosition()` throws an exception. This thrown exception, however, will be eliminated in a follow-up PR (#5320) that handles the not-on-dragway scenario. Would you be OK not having this unit test coverage given this follow-up PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jadecastro, I replied to your comments. PTAL. Thanks!

---

Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 297 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke24-lmPUOU42u3BlGE:-Ke3nKY5cUNsvJWOUEDM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L297)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, consider replacing the limits on `x`, `y`, and `z` with their named variables `length_`, `lane_width_`, etc.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke3nR6k78x7POd6c6CE:b-v85mw9) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, there are over 65,000 points checked in these two unit tests alone. Is the entire grid necessary? Or is it possible to pare this down to a handful of interesting points you'd like to check?
</blockquote></details>

I figured computation is cheap, as indicated by the fact that this unit test is still "small" from an execution-time perspective.

```
$ bazel test //drake/automotive/maliput/dragway:dragway_test
INFO: Found 1 test target...
Target //drake/automotive/maliput/dragway:dragway_test up-to-date:
  bazel-bin/drake/automotive/maliput/dragway/dragway_test
INFO: Elapsed time: 0.787s, Critical Path: 0.10s
//drake/automotive/maliput/dragway:dragway_test                          PASSED in 0.1s

Executed 1 out of 1 test: 1 test passes.
```

I could increase the loop increment size. What do you think about doubling them?

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 320 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke24ax6NevJzPfBPDhu:-Ke3oD3k_WfAUinZw3tM:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L320)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, consider replacing the limits on `x`, `y`, and `z` with their named variables `length_`, `lane_width_`, etc.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 338 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke22ndSOQ9uWGxnOeuD:-Ke3o1ype6CCaAmpP_0F:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L338)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

BTW, consider replacing `3`  with `lane_width / 2`.  Ditto below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-Ke3oWxb0BuGNc3Wq0FP:b-a3b6u0)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke1TwrtvZZ55Glk52sQ:-Ke4Ftq2rqSSjO7g0DjY:b7zfzbp) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/road_geometry.cc#L52)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Since this is a private method, the only way to unit test this is to check whether `ToRoadPosition()` throws an exception. This thrown exception, however, will be eliminated in a follow-up PR (#5320) that handles the not-on-dragway scenario. Would you be OK not having this unit test coverage given this follow-up PR?
</blockquote></details>

Then, perhaps the resolution is to add a comment in the unit test stating that it also verifies that `IsGeoPositionOnDragway` doesn't incorrectly throw an exception.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4BhCAdNvMoi-PkNQp:bls8pdj) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I figured computation is cheap, as indicated by the fact that this unit test is still "small" from an execution-time perspective.

```
$ bazel test //drake/automotive/maliput/dragway:dragway_test
INFO: Found 1 test target...
Target //drake/automotive/maliput/dragway:dragway_test up-to-date:
  bazel-bin/drake/automotive/maliput/dragway/dragway_test
INFO: Elapsed time: 0.787s, Critical Path: 0.10s
//drake/automotive/maliput/dragway:dragway_test                          PASSED in 0.1s

Executed 1 out of 1 test: 1 test passes.
```

I could increase the loop increment size. What do you think about doubling them?
</blockquote></details>

Regardless of runtime, I'm wondering whether there's any value adding points beyond just a grid including the four corners, and perhaps also four edge points, plus one interior point. 

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 322 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke4ANY-hhHEOo_yFFwN:-Ke4ANY0d58iBhf_gCmV:b-212twv) ([raw file](https://github.com/robotlocomotion/drake/blob/42d4260944cac85d44385fd26ffd57a8049a58f8/drake/automotive/maliput/dragway/test/dragway_test.cc#L322)):*
> ```cpp
>   for (double x = 0; x <= length_; x += 0.5) {
>     for (double y = 0; y <= lane_width_ + shoulder_width_; y += 0.1) {
>       for (double z = 0; z <= 10; z += 2.5) {
> ```

I am wondering whether the `z`-direction tests are even worthwhile, as `IsGeoPositionOnRoadway()` ignores `z` information (you have documentation to that effect too).  Ditto for the above unit test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/dragway/road_geometry.cc, line 52 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke1TwrtvZZ55Glk52sQ:-Ke4HxtAjwDfT3LWr-5K:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/road_geometry.cc#L52)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Then, perhaps the resolution is to add a comment in the unit test stating that it also verifies that `IsGeoPositionOnDragway` doesn't incorrectly throw an exception.
</blockquote></details>

Done.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4I0Suwnt-u67eqmc6:bh5bj2t) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Regardless of runtime, I'm wondering whether there's any value adding points beyond just a grid including the four corners, and perhaps also four edge points, plus one interior point. 
</blockquote></details>

Good question! Since I know the implementation details, I'm fairly confident the answer is "no". However, there is possibility that future developers may come along and "optimize" the implementation such that the answer becomes "yes". Should we protect ourselves from such future "optimization"?

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 322 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke4ANY-hhHEOo_yFFwN:-Ke4IZAA4vY37wWe1csk:b-cx4vcv) ([raw file](https://github.com/robotlocomotion/drake/blob/42d4260944cac85d44385fd26ffd57a8049a58f8/drake/automotive/maliput/dragway/test/dragway_test.cc#L322)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

I am wondering whether the `z`-direction tests are even worthwhile, as `IsGeoPositionOnRoadway()` ignores `z` information (you have documentation to that effect too).  Ditto for the above unit test.
</blockquote></details>

This is to "lock in" the expected semantic that the returned `RoadPosition` is always at `h = 0` regardless of what `z` is provided. I just realized that an assertion should be included in `RoadGeometry::DoToRoadPosition()` that double checks that `h = 0`. I added this assertion. However, even with this assertion, I believe it's still best practice to spot check that `h = 0` even when various non-zero `z` values are provided. Thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4LnltZvmM5AX5Phs7:bik6dw) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Good question! Since I know the implementation details, I'm fairly confident the answer is "no". However, there is possibility that future developers may come along and "optimize" the implementation such that the answer becomes "yes". Should we protect ourselves from such future "optimization"?
</blockquote></details>

Discussions posted by a reviewer carry a presumption of validity; the burden of proof to ignore a defect is on the author, not on the reviewer to demonstrate its validity.

In this case, unless the author has a convincing argument for either "the breadth of this test is likely to catch a future defect" or "writing it as a loop is clearer or briefer than just testing a few chosen points", then the review comment that "this code serves no purpose" should carry the day, and the test should be reworked to satisfy the discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4N1Rmp9ep2GGaDPMW:b-6dczaq) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Discussions posted by a reviewer carry a presumption of validity; the burden of proof to ignore a defect is on the author, not on the reviewer to demonstrate its validity.

In this case, unless the author has a convincing argument for either "the breadth of this test is likely to catch a future defect" or "writing it as a loop is clearer or briefer than just testing a few chosen points", then the review comment that "this code serves no purpose" should carry the day, and the test should be reworked to satisfy the discussion.
</blockquote></details>

@jadecastro, please let me know if you think my arguments above are convincing. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4N9k1bEItbK1tqaNj:bdqsao0) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

@jadecastro, please let me know if you think my arguments above are convincing. Thanks!
</blockquote></details>

There was no argument posted above.  A strawman "someone might actively try to make the DUT be stupid" is not an argument.  Please cite a specific implementation defect that is likely to occur, that would not be identified by a few-chosen-points test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 322 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke4ANY-hhHEOo_yFFwN:-Ke4Mr6Ev5vJFlf5dhN1:b-8tk1h3) ([raw file](https://github.com/robotlocomotion/drake/blob/42d4260944cac85d44385fd26ffd57a8049a58f8/drake/automotive/maliput/dragway/test/dragway_test.cc#L322)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This is to "lock in" the expected semantic that the returned `RoadPosition` is always at `h = 0` regardless of what `z` is provided. I just realized that an assertion should be included in `RoadGeometry::DoToRoadPosition()` that double checks that `h = 0`. I added this assertion. However, even with this assertion, I believe it's still best practice to spot check that `h = 0` even when various non-zero `z` values are provided. Thoughts?
</blockquote></details>

Nothing in the current code makes that intent clear.  Better than looping over `z` would be a single-point (one-liner) case with a comment beforehand `// Non-zero z are still mapped to zero h.` at which point the intent is clear, and we can decide if the mapping is what we want.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4PddNFZX-LzxWsPHx:b5j0hu2) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

My argument is someone may decide to optimize the implementation in ways that result in a few spot checks failing to detect the error. Whether this is "actively try to make the DUT be stupid" is subjective.
</blockquote></details>

That argument also presumes that such a PR doesn't add any new tests to cover the new code paths, which seems unlikely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4OFVBqvOfU4gTYcbr:b-5rco0f) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

That argument also presumes that such a PR doesn't add any new tests to cover the new code paths, which seems unlikely.
</blockquote></details>

Let me try to ask a slightly different question. Is the dragway _always_ a linear road? The answer might help shed light on whether or not there is a counterexample to the handful-of-points question, and the correctness of any implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4RAWJ-isSXf_5CX1g:bksnz6s) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

Let me try to ask a slightly different question. Is the dragway _always_ a linear road? The answer might help shed light on whether or not there is a counterexample to the handful-of-points question, and the correctness of any implementation.
</blockquote></details>

It is always an axis-aligned rectangle with straight lanes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4Ry4frHH0RTqyo1g1:b9gkdx7) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It is always an axis-aligned rectangle with straight lanes.
</blockquote></details>

Yes, that's correct. Based on that, do you feel safe decreasing the rigorousness of this unit test by spot checking fewer points?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 322 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke4ANY-hhHEOo_yFFwN:-Ke4Rz0gqAXKOTtGeFfs:b2vatms) ([raw file](https://github.com/robotlocomotion/drake/blob/42d4260944cac85d44385fd26ffd57a8049a58f8/drake/automotive/maliput/dragway/test/dragway_test.cc#L322)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Nothing in the current code makes that intent clear.  Better than looping over `z` would be a single-point (one-liner) case with a comment beforehand `// Non-zero z are still mapped to zero h.` at which point the intent is clear, and we can decide if the mapping is what we want.
</blockquote></details>

... and reading the maliput API docs, it seems to me that `h` should be non-zero when `z` is non-zero.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4WES1iYXappNA4V7U:b-s1idix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

As Jeremy mentions, I think the object is really for you to state a possible counterexample implementation to justify the need for more points. 

My argument is that your unit test is one that diagnoses false positives in your implementation, given the ground truth that the road is an axis-aligned rectangle. On these grounds, it suffices that checking your function at extreme points should be sufficient and no other more computation than that is necessary. 

However, the possibility that others can re-implement the function is fair, and you should describe, then, what possible implementations would cause the four-corners test to be unsound.
</blockquote></details>

Interesting.  I had not considered the possibility that my current test could diagnose false positives. Is that really true?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4Xf3CW5Ne4hgZ3pdl:b-bukig4) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Interesting.  I had not considered the possibility that my current test could diagnose false positives. Is that really true?
</blockquote></details>

to clarify, for instance: returning the incorrect result that `road_position.lane` is not same as `expected_lane`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 322 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke4ANY-hhHEOo_yFFwN:-Ke4nQLA4RtP5Er08DFq:b-cz6gez) ([raw file](https://github.com/robotlocomotion/drake/blob/42d4260944cac85d44385fd26ffd57a8049a58f8/drake/automotive/maliput/dragway/test/dragway_test.cc#L322)):*
Hmm, this is the contract that I'm looking at:

http://drake.mit.edu/doxygen_cxx/classdrake_1_1maliput_1_1api_1_1_road_geometry.html#a3e83bdf28024a8b67a948507b3e4a936

Specifically, this line:

> The return value is the RoadPosition of the point in the RoadGeometry's manifold which is, in the world frame, closest to geo_position.

So I guess the question boils down to: what is the `RoadGeometry`'s manifold? In my mind, the manifold for the dragway is the rectangle formed by the `RoadGeometry`'s driveable region at `z == 0`. Do you instead picture the manifold to be the hyperrectable of infinite height above the driveable region?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 322 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke4ANY-hhHEOo_yFFwN:-Ke4oImGa0mLv59Uer8S:bcj6dv8) ([raw file](https://github.com/robotlocomotion/drake/blob/42d4260944cac85d44385fd26ffd57a8049a58f8/drake/automotive/maliput/dragway/test/dragway_test.cc#L322)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Hmm, this is the contract that I'm looking at:

http://drake.mit.edu/doxygen_cxx/classdrake_1_1maliput_1_1api_1_1_road_geometry.html#a3e83bdf28024a8b67a948507b3e4a936

Specifically, this line:

> The return value is the RoadPosition of the point in the RoadGeometry's manifold which is, in the world frame, closest to geo_position.

So I guess the question boils down to: what is the `RoadGeometry`'s manifold? In my mind, the manifold for the dragway is the rectangle formed by the `RoadGeometry`'s driveable region at `z == 0`. Do you instead picture the manifold to be the hyperrectable of infinite height above the driveable region?
</blockquote></details>

Yes, given the other maliput API comments, I believe the intent is to extend the volume to sweep all values of `h`.  Specifically, at that link it says "If geo_position is contained within the volumes of multiple Segments, then ToRoadPosition() will choose a Segment which yields the minimum height h value in the result." which would make no sense if all h's for forever zero in this context.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 322 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke4ANY-hhHEOo_yFFwN:-Ke4oiid7jkCV7d2WCBT:b-fobxn7) ([raw file](https://github.com/robotlocomotion/drake/blob/42d4260944cac85d44385fd26ffd57a8049a58f8/drake/automotive/maliput/dragway/test/dragway_test.cc#L322)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yes, given the other maliput API comments, I believe the intent is to extend the volume to sweep all values of `h`.  Specifically, at that link it says "If geo_position is contained within the volumes of multiple Segments, then ToRoadPosition() will choose a Segment which yields the minimum height h value in the result." which would make no sense if all h's for forever zero in this context.
</blockquote></details>

Alright, I will update this PR to select an `h` that matches the supplied `z`. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4vleJitMDz26ZUggH:bg405cr) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

to clarify, for instance: returning the incorrect result that `road_position.lane` is not same as `expected_lane`.
</blockquote></details>

I'm not sure I follow. Can you explain?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke4x9-OOfkcTEARYIeq:b-fgltrg) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I'm not sure I follow. Can you explain?
</blockquote></details>

For now, please just strike "false positives in" from the above comment, and proceed from there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@jwnimmer-tri and @jadecastro, I addressed your comments. PTAL. Thanks.

---

Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/maliput/dragway/test/dragway_test.cc, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-Ke27PFH2b8f03TMF0Ds:-Ke69GLjgjXtodwx0jVA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0d8ca81c69d4f23a0340c18b4cd5b8e1d51346b7/drake/automotive/maliput/dragway/test/dragway_test.cc#L299)):*
<details><summary><i>Previously, jadecastro (Jonathan DeCastro) wrote…</i></summary><blockquote>

For now, please just strike "false positives in" from the above comment, and proceed from there.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-Ke69HiMG5O_8ma4uw9c:b-ya8z5q)*
<!-- Sent from Reviewable.io -->

jadecastro commented:
Thanks @liangfok! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 2 of 2 files at r6.
Review status: 2 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-Ke6AF51YdQ1I50_YQR5:b-9hoxxa)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Thanks! Now waiting on CI.

---

Review status: 1 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5313#-:-Ke6Aj4BBRlymgA7-9Qg:b-5y5ebz)*
<!-- Sent from Reviewable.io -->
